src/targets/gpu/device/layernorm.cpp

```diff
@@ -0,0 +1,64 @@
+#include <migraphx/gpu/device/layernorm.hpp>
+#include <migraphx/gpu/device/reduce.hpp>
+#include <migraphx/gpu/device/pow.hpp>
+#include <migraphx/gpu/device/fast_div.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+namespace device {
+
+// m = x - mean(x)
+// m / sqrt(mean(m ^ 2) + 1e-12)
+void layernorm(hipStream_t stream, const argument& result, const argument& arg1)
+{
+    auto relements    = arg1.get_shape().lens().back();
+    auto nelements    = result.get_shape().elements() / relements;
+    auto input_shape  = arg1.get_shape();
+    auto output_shape = result.get_shape();
+    auto reduce_output_lens(output_shape.lens());
+    reduce_output_lens.back() = 1;
+
+    std::vector<index_int> reduce_lens = get_reduce_lens(input_shape.lens(), reduce_output_lens);
+
+    hip_visit_all(result, arg1)([&](auto output, auto input) {
+        using value_type = typename decltype(input)::value_type;
+
+        const std::size_t max_block_size = 256;
+        const std::size_t block_size     = compute_block_size(relements, max_block_size);
+        const std::size_t block_size_div = encode_divisor(block_size);
+
+        gs_launch(stream, nelements * block_size, block_size)([=](auto i, auto idx) __device__ {
+            const auto out_idx  = i / block_size;
+            const auto base_idx = out_idx * relements;
+            value_type x_data[4];
```

Comment:
Should we initialize x_data to 0s?

Reply:
There may be a performance cost initializing it.

Reply:
I will leave it as is for now, unless there are some correctness issues in the future